README.md
# azure-log-analytics-data-collector Ruby Client

[Azure Log Analytics Data Collector API](https://docs.microsoft.com/en-us/azure/azure-monitor/platform/data-collector-api) Client Libraries for Ruby. The repository was originally created for multiple programming languages, but it was refactored as a dedicated one for Ruby client. Python and PHP client libraries were moved to [azure-log-analytics-data-colloector-python](https://github.com/yokawasa/azure-log-analytics-data-collector-python) and [azure-log-analytics-data-colloector-php](https://github.com/yokawasa/azure-log-analytics-data-collector-php) respectively.


## Retry policy

The client internal leverage [rest-client] to send HTTP request to the API. The client library retries request using the rest-client on the following status code (which is [recommended action](https://docs.microsoft.com/en-us/azure/azure-monitor/platform/data-collector-api)).
  * `429` - Too Many Requests
  * `500` - Internal Server Error
  * `503` -	Service Unavailable

By default, the client library retres for a total of `3` times, sleeping `5 sec` between retries. The number of retries and sleeping time between retries can be changed with `set_retries` in the client class.

```ruby
  def set_retries(max_retries, retry_sleep_period)
    @max_retries = max_retries
    @retry_sleep_period = retry_sleep_period
  end
```

## Installation
```bash
gem install azure-loganalytics-datacollector-api
```

## Sample code (Ruby Client)
### Sample1 - No time_generated_field option
```ruby
require "azure/loganalytics/datacollectorapi/client"

customer_id = '<Customer ID aka WorkspaceID String>'
shared_key = '<The primary or the secondary Connected Sources client authentication key>'
log_type = "MyCustomLog"

posting_records = []
record1= {
  :string => "MyText1",
  :boolean => true,
  :number => 100
}
record2= {
  :string => "MyText2",
  :boolean => false,
  :number => 200
}
posting_records.push(record1)
posting_records.push(record2)

client=Azure::Loganalytics::Datacollectorapi::Client::new( customer_id, shared_key)
res = client.post_data(log_type, posting_records)
puts res
puts "res code=#{res.code}"

if Azure::Loganalytics::Datacollectorapi::Client.is_success(res)
  puts "operation was succeeded!"
else
  puts "operation was failured!"
end
```

### Sample2 - With time_generated_field option
```ruby
require "azure/loganalytics/datacollectorapi/client"

customer_id = '<Customer ID aka WorkspaceID String>'
shared_key = '<The primary or the secondary Connected Sources client authentication key>'
log_type = "MyCustomLog"

posting_records = []
record1= {
  :string => "MyText1",
  :boolean => true,
  :number => 100,
  :timegen => "2017-11-23T11:13:35.576Z" # YYYY-MM-DDThh:mm:ssZ
}
record2= {
  :string => "MyText2",
  :boolean => false,
  :number => 200,
  :timegen => "2017-11-23T12:13:35.576Z" # YYYY-MM-DDThh:mm:ssZ
}
posting_records.push(record1)
posting_records.push(record2)

time_generated_field = "timegen"
client=Azure::Loganalytics::Datacollectorapi::Client::new( customer_id, shared_key)
res = client.post_data(log_type, posting_records, time_generated_field)
puts res
puts "res code=#{res.code}"

if Azure::Loganalytics::Datacollectorapi::Client.is_success(res)
  puts "operation was succeeded!"
else
  puts "operation was failured!"
end
```

### Sample3 - With time_generated_field and azure_resource_id option
Supported setting azure_resource_id option from version [0.2.0](https://github.com/yokawasa/azure-log-analytics-data-collector/releases/tag/v0.2.0)
```ruby
require "azure/loganalytics/datacollectorapi/client"

customer_id = '<Customer ID aka WorkspaceID String>'
shared_key = '<The primary or the secondary Connected Sources client authentication key>'
log_type = "MyCustomLog"

posting_records = []
record1= {
  :string => "MyText1",
  :boolean => true,
  :number => 100,
  :timegen => "2017-11-23T11:13:35.576Z" # YYYY-MM-DDThh:mm:ssZ
}
record2= {
  :string => "MyText2",
  :boolean => false,
  :number => 200,
  :timegen => "2017-11-23T12:13:35.576Z" # YYYY-MM-DDThh:mm:ssZ
}
posting_records.push(record1)
posting_records.push(record2)

time_generated_field = "timegen"

# Azure Resource ID
# [Azure Resource ID Format]
# /subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/{resourceProviderNamespace}/{resourceType}/{resourceName}
azure_resource_id ="/subscriptions/11111111-1111-1111-1111-111111111111/resourceGroups/otherResourceGroup/providers/Microsoft.Storage/storageAccounts/examplestorage"

client=Azure::Loganalytics::Datacollectorapi::Client::new( customer_id, shared_key)
res = client.post_data(log_type, posting_records, time_generated_field, azure_resource_id)
puts res
puts "res code=#{res.code}"

if Azure::Loganalytics::Datacollectorapi::Client.is_success(res)
  puts "operation was succeeded!"
else
  puts "operation was failured!"
end
```

### Sample4 - use proxy to access the API
```ruby
require "azure/loganalytics/datacollectorapi/client"

...
client=Azure::Loganalytics::Datacollectorapi::Client::new( customer_id, shared_key)
# client.set_proxy() # ENV['http_proxy'] is set by default 
client.set_proxy(your_proxy)
res = client.post_data(log_type, posting_records, time_generated_field)
puts res
...
```


## Change log

* [Changelog](ChangeLog.md)

## Links

* https://rubygems.org/gems/azure-loganalytics-datacollector-api
* [Azure Log Analytics Data Collecotr API](https://docs.microsoft.com/en-us/azure/log-analytics/log-analytics-data-collector-api)

## Development

After checking out the repo, run `bin/setup` to install dependencies. Then, run `rake spec` to run the tests. You can also run `bin/console` for an interactive prompt that will allow you to experiment.

To install this gem onto your local machine, run `bundle exec rake install`. To release a new version, update the version number in `version.rb`, and then run `bundle exec rake release`, which will create a git tag for the version, push git commits and tags, and push the `.gem` file to [rubygems.org](https://rubygems.org).

## Contributing

Bug reports and pull requests are welcome on GitHub at https://github.com/yokawasa/azure-log-analytics-data-collector. This project is intended to be a safe, welcoming space for collaboration, and contributors are expected to adhere to the [Contributor Covenant](http://contributor-covenant.org) code of conduct.

## License

The gem is available as open source under the terms of the [MIT License](http://opensource.org/licenses/MIT).


ChangeLog.md
# ChangeLog

## 0.5.0

* Check body size not to exceed data limit of 30MB in Azure Monitor Data Collection API - [issue #12](https://github.com/yokawasa/azure-log-analytics-data-collector/issues/12)

## 0.4.0

* restclient retries request on the following status code - [issue #10](https://github.com/yokawasa/azure-log-analytics-data-collector/issues/10)
  * 429 - Too Many Requests
  * 500 - Internal Server Error
  * 503 -	Service Unavailable

## 0.3.0

* Enhance log type validation: check not only alpha but also numeric, underscore, and character length (may not exceed 100) - [issue #11](https://github.com/yokawasa/azure-log-analytics-data-collector/issues/11)

## 0.2.0

* Support setting the x-ms-AzureResourceId Header - [issue #8](https://github.com/yokawasa/azure-log-analytics-data-collector/issues/8)

## 0.1.6

* fix CVE-2020-8130 - [issue #7](https://github.com/yokawasa/azure-log-analytics-data-collector/issues/7)

## 0.1.5

* Add endpoint parameter with the default value for Azure public - [PR#5](https://github.com/yokawasa/azure-log-analytics-data-collector/pull/5)

## 0.1.4

* Add `set_proxy` method to client - [issue#3](https://github.com/yokawasa/azure-log-analytics-data-collector/issues/3)

## 0.1.2

* fixedup bug - [issue #1](https://github.com/yokawasa/azure-log-analytics-data-collector/issues/1)

## 0.1.1

* rm helper.rb and include helper functions in client.rb

## 0.1.0

* Inital Release


lib/azure/loganalytics/datacollectorapi/client.rb
require "azure/loganalytics/datacollectorapi/version"

module Azure
  module Loganalytics
    module Datacollectorapi

      API_VERSION = '2016-04-01'.freeze

      class Client

        DEFAUT_MAX_RETRIES = 3.freeze
        DEFAULT_RETRY_SLEEP_PERIOD = 5.freeze
        # API Data Limits
        # https://docs.microsoft.com/en-us/azure/azure-monitor/platform/data-collector-api#data-limits
        MAX_BODY_BYTE_SIZE = 31457280.freeze   # 30 MB

        def initialize (customer_id, shared_key, endpoint ='ods.opinsights.azure.com')
          require 'rest-client'
          require 'json'
          require 'openssl'
          require 'base64'
          require 'time'

          @customer_id = customer_id
          @shared_key = shared_key
          @endpoint = endpoint
          @default_azure_resource_id = ''

          @max_retries = DEFAUT_MAX_RETRIES
          @retry_sleep_period = DEFAULT_RETRY_SLEEP_PERIOD    
        end

        def post_data(log_type, json_records, record_timestamp ='', azure_resource_id ='' )
          raise ConfigError, 'no log_type' if log_type.empty?
          raise ConfigError, 'log_type must only contain alpha numeric and _, and not exceed 100 chars' if not is_valid_log_type(log_type)
          raise ConfigError, 'no json_records' if json_records.empty?

          body =  json_records.to_json
          body_size = body.bytesize
          raise "too large payload (#{body_size})! max post data size is #{MAX_BODY_BYTE_SIZE} bytes!" if body_size >= MAX_BODY_BYTE_SIZE

          uri = sprintf("https://%s.%s/api/logs?api-version=%s",
                        @customer_id, @endpoint, API_VERSION)
          date = rfc1123date()
          sig = signature(date, body.bytesize)

          headers = {
              'Content-Type' => 'application/json',
              'Authorization' => sig,
              'Log-Type' => log_type,
              'x-ms-date' => date,
              'x-ms-AzureResourceId' => azure_resource_id.empty? ? @default_azure_resource_id : azure_resource_id,
              'time-generated-field' => record_timestamp
          }

          retries = 0 
          begin
            res = RestClient.post( uri, body, headers)
            res
          rescue => e
            c = e.response.code.to_i
            if c == 429 || c == 500 || c==503
              if retries < @max_retries
                retries += 1
                sleep(@retry_sleep_period)
                retry 
              end
            end
            raise e
          end
        end

        def set_proxy(proxy='')
          RestClient.proxy = proxy.empty? ? ENV['http_proxy'] : proxy
        end
        
        def set_default_azure_resoruce_id(azure_resource_id)
          @default_azure_resource_id = azure_resource_id
        end
        
        def set_retres(max_retries, retry_sleep_period)
          @max_retries = max_retries
          @retry_sleep_period = retry_sleep_period
        end

        def self.is_success(res)
          return (res.code == 200) ? true : false
        end

        private

        def is_valid_log_type(s)
          return ( s.match(/^[a-zA-Z0-9_]+$/) && s.length <= 100 ) ? true : false
        end

        def rfc1123date()
          t = Time.now
          t.httpdate()
        end

        def signature(date, content_length)
          sigs = sprintf("POST\n%d\napplication/json\nx-ms-date:%s\n/api/logs",
                        content_length, date)
          utf8_sigs = sigs.encode('utf-8')
          decoded_shared_key = Base64.decode64(@shared_key)
          hmac_sha256_sigs = OpenSSL::HMAC.digest(OpenSSL::Digest.new('sha256'), decoded_shared_key, utf8_sigs)
          encoded_hash = Base64.encode64(hmac_sha256_sigs)
          authorization = sprintf("SharedKey %s:%s", @customer_id,encoded_hash)
          authorization
        end

      end # end of class

    end
  end
end
